### PR TITLE
Do not export per user and integration Alertmanager metrics when value is 0

### PR DESCRIPTION
**Pull request by pracucci**
#### What this PR does
As discussed in #1750, in this PR I propose to not export Alertmanager counter metrics having both `user` and `integration` labels when their value is 0.

Currently these metrics are exposed for any integration, even if an user doesn't use that integration. Alertmanager supports 10 receiver integrations, and this leads to high cardinality metrics when Alertmanager is handling a large number of tenants. For example, if you run 3 Alertmanager replicas, you have 10k tenants, and each tenant configures 2 integrations, before the change we have in total `3 * 10k * 10 = 300k` metrics, while after the change we have `3 * 10k * 2 = 60k` metrics.

#### Which issue(s) this PR fixes or relates to

Part of #1750

#### Checklist

- [ ] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/util/metrics_helper.go [hunk 1]
```diff
@@ -165,13 +169,20 @@ func (d MetricFamiliesPerUser) SendSumOfCountersPerUser(out chan<- prometheus.Me
 // SendSumOfCountersPerUserWithLabels provides metrics with the provided label names on a per-user basis. This function assumes that `user` is the
 // first label on the provided metric Desc
 func (d MetricFamiliesPerUser) SendSumOfCountersPerUserWithLabels(out chan<- prometheus.Metric, desc *prometheus.Desc, metric string, labelNames ...string) {
+	d.SendSumOfCountersPerUserWithLabelsAndOptions(out, desc, metric, labelNames)
```
Comment (gotjosh) [inline pkg/util/metrics_helper.go:172]: 💯  on not having to do a shotgun surgery ⭐

**Review by gotjosh — APPROVED** [review]
LGTM.

If the user does have the integration configured but has not received any notifications it might be misleading - I don't see a good way to preserve that at a glance and the positives overcome the negatives by a mile.

⭐

**Review by bboreham — APPROVED** [review]
Seems legit.

File: CHANGELOG.md [hunk 2]
```diff
@@ -13,6 +13,12 @@
 * [CHANGE] Store-gateway: enabled attributes in-memory cache by default. New default configuration is `-blocks-storage.bucket-store.chunks-cache.attributes-in-memory-max-items=50000`. #1727
 * [CHANGE] Compactor: Removed the metric `cortex_compactor_garbage_collected_blocks_total` since it duplicates `cortex_compactor_blocks_marked_for_deletion_total`. #1728
 * [CHANGE] All: Logs that used the`org_id` label now use `user` label. #1634 #1758
+* [CHANGE] Alertmanager: the following metrics are not exported for a given `user` and `integration` when the metric value is zero: #1783
+  * `cortex_alertmanager_notifications_total`
+  * `cortex_alertmanager_notifications_failed_total`
+  * `cortex_alertmanager_notification_requests_total`
+  * `cortex_alertmanager_notification_requests_failed_total`
+  * `cortex_alertmanager_notification_rate_limited_total`
```
[the remark this answers is not in the document]
Comment (pracucci) [reply CHANGELOG.md:21]: We discussed it and we won't rename it until we can't provide a smooth upgrade path to users.

**Comment by pstibrany** [comment]
To make all checks pass again, please rebase this PR on top of `main`. Thank you.

File: pkg/alertmanager/alertmanager_metrics.go [hunk 3]
```diff
@@ -296,10 +296,10 @@ func (m *alertmanagerMetrics) Collect(out chan<- prometheus.Metric) {
 	data.SendSumOfCountersPerUser(out, m.alertsReceived, "alertmanager_alerts_received_total")
 	data.SendSumOfCountersPerUser(out, m.alertsInvalid, "alertmanager_alerts_invalid_total")
 
-	data.SendSumOfCountersPerUserWithLabels(out, m.numNotifications, "alertmanager_notifications_total", "integration")
-	data.SendSumOfCountersPerUserWithLabels(out, m.numFailedNotifications, "alertmanager_notifications_failed_total", "integration")
-	data.SendSumOfCountersPerUserWithLabels(out, m.numNotificationRequestsTotal, "alertmanager_notification_requests_total", "integration")
-	data.SendSumOfCountersPerUserWithLabels(out, m.numNotificationRequestsFailedTotal, "alertmanager_notification_requests_failed_total", "integration")
+	data.SendSumOfCountersPerUserWithLabelsAndOptions(out, m.numNotifications, "alertmanager_notifications_total", []string{"integration"}, util.SkipZeroValueMetrics)
```
Comment (pstibrany) [inline pkg/alertmanager/alertmanager_metrics.go:299]: On first sight "AndOptions" looked like premature generalization to me, and I was going to suggest to generalize only when there are at least three variants in the future.

But there are already 6! :)

```
SendSumOfCounters
SendSumOfCountersPerUser
SendSumOfCountersPerUserWithLabels
SendSumOfCountersPerUserWithLabelsAndOptions
SendSumOfCountersWithLabels
```

Perhaps we should only have `SendSumOfCounters` and variants should really be "options"?

(But that's for another PR)

**Review by pstibrany — APPROVED** [review]
Thank you.

File: pkg/alertmanager/alertmanager_metrics.go (hunk 3 above)
Comment (pracucci) [reply pkg/alertmanager/alertmanager_metrics.go:299]: _Looking into it._

**Pull request merged**